elections/models.py

```diff
@@ -40,37 +43,38 @@ class Election(models.Model):
     vote_end = models.DateTimeField(help_text="End of voting")
     allowed_voters = models.ManyToManyField(User, through="Voter")
 
+    def get_results(self):
+        return {ballot.description: ballot.get_results() for ballot in self.ballots.all()}
+
+    def get_nominees(self):
+        return {ballot.description: ballot.get_nominees() for ballot in self.ballots.all()}
+
     class Meta:
         ordering = ("-vote_start",)
 
     def __unicode__(self):
         return unicode(self.name)
 
-class Voter(models.Model):
+class Voter(AbstractToken):
     """We can store a token to send a unique email to users so we don't require login.
 
     Once used, this token should be removed to avoid re-use if so desired.
     """
     user = models.ForeignKey(User, on_delete=models.CASCADE)
     election = models.ForeignKey("Election", on_delete=models.CASCADE)
-    token = models.CharField(max_length=10) # base64, is more than enough
-
-    def save(self, *args, **kwargs):
-        if not self.token:
-            self.token = get_random_string(length=10)
-        super(Voter, self).save(*args, **kwargs)
+    has_voted = models.BooleanField(default=False)
 
     class Meta:
         # There should be no duplicate tokens (or users) in an election!
-        unique_together=(("election", "token"), ("election", "user"))
+        unique_together=(("election", "user"), )
```

Comment:
this is because each user must already have a unique token?

Reply:
Tokens have a unique constraint on their own now, so no need to have a weaker version of it here.